code/fripp.c

```diff
@@ -0,0 +1,139 @@
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <flux/core.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#define FRIPP_MAX_PACKET_LEN 1440
+#define FRIPP_INTERNAL_BUF_SIZE (FRIPP_MAX_PACKET_LEN * 5)
+
+struct fripp_ctx {
+    flux_t *h;
+    struct sockaddr_in si_server;
+};
+
+/* Bind a socket address to hostname:port.
+ */
+int fripp_bind_server (struct fripp_ctx *ctx, const char *hostname, uint16_t port)
+{
+    memset (&ctx->si_server, 0, sizeof (ctx->si_server));
+    ctx->si_server.sin_family = AF_INET;
+    ctx->si_server.sin_port = htons (port);
+
+    if (!inet_aton (hostname, &ctx->si_server.sin_addr)) {
+        flux_log_error (ctx->h, "error creating server address");
+        return -1;
+    }
+
+    return 0;
+}
+
+/* Split a packet that is longer than FRIPP_MAX_PACKET_LEN
+ * by the last '\n' character, and replace it with a '\0'.
+ */
+static int split_packet (char *packet)
+{
+    for (int i = FRIPP_MAX_PACKET_LEN - 1; i >= 0; i--) {
+        if (packet[i] == '\n' || packet[i] == '\0') {
+            packet[i] = '\0';
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+
+/* Send the 'packet' string in one or more udp packets to
+ * the bound server within 'ctx'.
+ */
+static int fripp_send_metrics (struct fripp_ctx *ctx, char *packet)
+{
+    int sock, sock_len = sizeof (ctx->si_server);
+
+    if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) == -1)
+        return -1;
+
+    int len = strlen (packet);
+    bool split = false;
+
+    if (len > FRIPP_MAX_PACKET_LEN) {
+        split = true;
+
+        if ((len = split_packet (packet)) == -1) {
+            close (sock);
+            return -1;
+        }
+    }
+
+    if (sendto (sock, packet, len, 0, (void *) &ctx->si_server, sock_len) < 0)
+        flux_log (ctx->h, LOG_ALERT, "packet %s dropped", packet);
+
+    close (sock);
+
+    if (split)
+        return fripp_send_metrics (ctx, &packet[len + 1]);
+
+    return 0;
+}
+
+void fripp_format_send (struct fripp_ctx *ctx, const char *fmt, ...)
+{
+    char *packet;
+    char buf[FRIPP_INTERNAL_BUF_SIZE];
+
+    va_list ap;
+    va_start (ap, fmt);
+
+    if (vasprintf (&packet, fmt, ap) < 0) {
+        (void) vsnprintf (buf, sizeof (buf), fmt, ap);
+        packet = buf;
+    }
+
+    fripp_send_metrics (ctx, packet);
+    va_end (ap);
+
+    if (packet != buf)
+        free (packet);
+}
+
+void fripp_ctx_destroy (struct fripp_ctx *ctx)
+{
+    free (ctx);
+}
+
+struct fripp_ctx *fripp_ctx_create (flux_t *h, const char *hostname, int port)
+{
+    struct fripp_ctx *ctx;
+
+    if (!(ctx = malloc (sizeof (*ctx)))) {
+        flux_log_error (h, "fripp_ctx_create");
+        return NULL;
+    }
+    if (fripp_bind_server (ctx, hostname, (uint16_t) port)) {
+        flux_log_error (h, "fripp_bind_server");
+        free (ctx);
+        ctx = NULL;
+    }
```

Comment:
Suggestion: just put the socket runes in here directly, and avoid the slightly confusing name "bind_server" name, which to me sounds like should contain a call to bind(2).